include/seqan3/range/container/small_vector.hpp

```diff
@@ -0,0 +1,931 @@
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    constexpr void push_back(value_type const value) noexcept
+    {
+        assert(sz < capacity_);
+        data_[sz] = value;
+        ++sz;
+    }
+
+    /*!\brief Removes the last element of the container.
+     *
+     * Calling pop_back() on an empty container is undefined. In debug mode an assertion will be thrown.
+     *
+     * No iterators or references except for back() and end() are invalidated.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    constexpr void pop_back() noexcept
+    {
+        assert(sz > 0);
+        --sz;
+    }
+
+    /*!\brief Resizes the container to contain count elements.
+     * \param[in] count The new size.
+     *
+     * If count is greater than capacity this is undefined behaviour.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    constexpr void resize(size_type const count) noexcept
+    {
+        assert(count <= capacity_);
+        sz = count;
+    }
+
+    /*!\copybrief resize()
+     * \param value Append copies of value when resizing.
+     * \copydetails resize()
+     */
+    constexpr void resize(size_type const count, value_type const value) noexcept
+    {
+        assert(count <= capacity_);
+        for (size_t i = sz; i < count; ++i)
+            data_[i] = value;
+        sz = count;
+    }
+
+    /*!\brief Swap contents with another instance.
+     * \param rhs The other instance to swap with.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of both containers.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee if value_type is std::is_nothrow_copy_constructible.
+     */
+    constexpr void swap(small_vector & rhs) noexcept(is_noexcept)
+    {
+        auto tmp = *this;
+
+        data_ = rhs.data_;
+        sz = rhs.sz;
+
+        rhs.data_ = tmp.data_;
+        rhs.sz = tmp.sz;
+    }
+
+    //!\copydoc swap()
+    constexpr void swap(small_vector && rhs) noexcept(is_noexcept)
+    {
+        data_ = rhs.data_;
+        sz = rhs.sz;
+    }
+    //!\}
+
+    /*!\brief Swap contents with another instance.
+     * \param lhs The first instance.
+     * \param rhs The other instance to swap with.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of both containers.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee if value_type is std::is_nothrow_copy_constructible.
+     */
+    friend constexpr void swap(small_vector & lhs, small_vector & rhs) noexcept(is_noexcept)
+    {
+        lhs.swap(rhs);
+    }
+
+    //!\overload
+    template <size_t capacity2>
+    friend constexpr void swap(small_vector && lhs, small_vector && rhs) noexcept(is_noexcept)
+    {
+        lhs.swap(rhs);
+    }
```

Comment:
Concerning the swaps, the `std::swap` free function should be auto-generated since you have move-cosntructors so you don't need the friends. In fact you could probably have the members functions delegate to the free function, too. (this will not perform the micro-optimisation of saving one copy for temporaries, though.

Reply:
> Concerning the swaps, the std::swap free function should be auto-generated since you have move-cosntructors so you don't need the friends.

those friends are not in the std namespace but in the seqan3 namespace? Don;t we have our own swap?

> In fact you could probably have the members functions delegate to the free function, too.

No I cannot because neither `std::swap` nor `ranges::swap` are constexpr (prior to c++20)